In [86]:
import pandas as pd
import numpy as np

In [87]:
import os
import tarfile
from six.moves import urllib

DOWNLOAD_ROOT = "https://raw.githubusercontent.com/ageron/handson-ml/master/"
HOUSING_PATH = os.path.join("datasets", "housing")
HOUSING_URL = DOWNLOAD_ROOT + "datasets/housing/housing.tgz"

def fetch_housing_data(housing_url=HOUSING_URL, housing_path=HOUSING_PATH):
    os.makedirs(housing_path, exist_ok=True)
    tgz_path = os.path.join(housing_path, "housing.tgz")
    urllib.request.urlretrieve(housing_url, tgz_path)
    housing_tgz = tarfile.open(tgz_path)
    housing_tgz.extractall(path=housing_path)
    housing_tgz.close()

In [88]:
fetch_housing_data()

In [89]:
def load_housing_data(housing_path=HOUSING_PATH):
    csv_path = os.path.join(housing_path, "housing.csv")
    return pd.read_csv(csv_path)

In [90]:
housing = load_housing_data()
housing.head()

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0    -122.23     37.88                41.0        880.0           129.0   
1    -122.22     37.86                21.0       7099.0          1106.0   
2    -122.24     37.85                52.0       1467.0           190.0   
3    -122.25     37.85                52.0       1274.0           235.0   
4    -122.25     37.85                52.0       1627.0           280.0   

   population  households  median_income  median_house_value ocean_proximity  
0       322.0       126.0         8.3252            452600.0        NEAR BAY  
1      2401.0      1138.0         8.3014            358500.0        NEAR BAY  
2       496.0       177.0         7.2574            352100.0        NEAR BAY  
3       558.0       219.0         5.6431            341300.0        NEAR BAY  
4       565.0       259.0         3.8462            342200.0        NEAR BAY

## Кодирование категориальных признаков

### Категориальные признаки

In [91]:
housing['ocean_proximity'].value_counts()

<1H OCEAN     9136
INLAND        6551
NEAR OCEAN    2658
NEAR BAY      2290
ISLAND           5
Name: ocean_proximity, dtype: int64

### Кодирование с помощью LabelEncoder

In [92]:
from sklearn.preprocessing import LabelEncoder

le = LabelEncoder()
cat_enc_le = le.fit_transform(housing['ocean_proximity'])
cat_enc_le

array([3, 3, 3, ..., 1, 1, 1])

In [93]:
np.unique(cat_enc_le)

array([0, 1, 2, 3, 4])

In [94]:
le.inverse_transform([0, 1, 2, 3, 4])

array(['<1H OCEAN', 'INLAND', 'ISLAND', 'NEAR BAY', 'NEAR OCEAN'],
      dtype=object)

In [95]:
housing['ocean_proximity_le'] = cat_enc_le
housing

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0        -122.23     37.88                41.0        880.0           129.0   
1        -122.22     37.86                21.0       7099.0          1106.0   
2        -122.24     37.85                52.0       1467.0           190.0   
3        -122.25     37.85                52.0       1274.0           235.0   
4        -122.25     37.85                52.0       1627.0           280.0   
...          ...       ...                 ...          ...             ...   
20635    -121.09     39.48                25.0       1665.0           374.0   
20636    -121.21     39.49                18.0        697.0           150.0   
20637    -121.22     39.43                17.0       2254.0           485.0   
20638    -121.32     39.43                18.0       1860.0           409.0   
20639    -121.24     39.37                16.0       2785.0           616.0   

       population  households  median_income  median_house_value  \
0           322.0       126.0         8.3252            452600.0   
1          2401.0      1138.0         8.3014            358500.0   
2           496.0       177.0         7.2574            352100.0   
3           558.0       219.0         5.6431            341300.0   
4           565.0       259.0         3.8462            342200.0   
...           ...         ...            ...                 ...   
20635       845.0       330.0         1.5603             78100.0   
20636       356.0       114.0         2.5568             77100.0   
20637      1007.0       433.0         1.7000             92300.0   
20638       741.0       349.0         1.8672             84700.0   
20639      1387.0       530.0         2.3886             89400.0   

      ocean_proximity  ocean_proximity_le  
0            NEAR BAY                   3  
1            NEAR BAY                   3  
2            NEAR BAY                   3  
3            NEAR BAY                   3  
4            NEAR BAY                   3  
...               ...                 ...  
20635          INLAND                   1  
20636          INLAND                   1  
20637          INLAND                   1  
20638          INLAND                   1  
20639          INLAND                   1  

[20640 rows x 11 columns]

### Кодирование с помощью OneHotEncoder

In [96]:
from sklearn.preprocessing import OneHotEncoder

ohe = OneHotEncoder()
cat_ohe = ohe.fit_transform(housing[['ocean_proximity']])

In [97]:
ohe.categories_

[array(['<1H OCEAN', 'INLAND', 'ISLAND', 'NEAR BAY', 'NEAR OCEAN'],
       dtype=object)]

In [98]:
cat_ohe.toarray()
#cat_ohe.A

array([[0., 0., 0., 1., 0.],
       [0., 0., 0., 1., 0.],
       [0., 0., 0., 1., 0.],
       ...,
       [0., 1., 0., 0., 0.],
       [0., 1., 0., 0., 0.],
       [0., 1., 0., 0., 0.]])

In [99]:
pd.get_dummies(housing['ocean_proximity'])

<1H OCEAN  INLAND  ISLAND  NEAR BAY  NEAR OCEAN
0              0       0       0         1           0
1              0       0       0         1           0
2              0       0       0         1           0
3              0       0       0         1           0
4              0       0       0         1           0
...          ...     ...     ...       ...         ...
20635          0       1       0         0           0
20636          0       1       0         0           0
20637          0       1       0         0           0
20638          0       1       0         0           0
20639          0       1       0         0           0

[20640 rows x 5 columns]

## Обработка пропусков в данных

### Первый способ определить признаки с нулевыми значениями
#### total_bedrooms имеет 20433 ненулевых объекта из 20640

In [100]:
housing.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20640 entries, 0 to 20639
Data columns (total 11 columns):
longitude             20640 non-null float64
latitude              20640 non-null float64
housing_median_age    20640 non-null float64
total_rooms           20640 non-null float64
total_bedrooms        20433 non-null float64
population            20640 non-null float64
households            20640 non-null float64
median_income         20640 non-null float64
median_house_value    20640 non-null float64
ocean_proximity       20640 non-null object
ocean_proximity_le    20640 non-null int32
dtypes: float64(9), int32(1), object(1)
memory usage: 1.7+ MB


#### Второй способ

In [101]:
housing.isnull().sum()

longitude               0
latitude                0
housing_median_age      0
total_rooms             0
total_bedrooms        207
population              0
households              0
median_income           0
median_house_value      0
ocean_proximity         0
ocean_proximity_le      0
dtype: int64

In [102]:
sample_incomplete_rows = housing[housing.isnull().any(axis=1)].head()
sample_incomplete_rows

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
290    -122.16     37.77                47.0       1256.0             NaN   
341    -122.17     37.75                38.0        992.0             NaN   
538    -122.28     37.78                29.0       5154.0             NaN   
563    -122.24     37.75                45.0        891.0             NaN   
696    -122.10     37.69                41.0        746.0             NaN   

     population  households  median_income  median_house_value  \
290       570.0       218.0         4.3750            161900.0   
341       732.0       259.0         1.6196             85100.0   
538      3741.0      1273.0         2.5762            173400.0   
563       384.0       146.0         4.9489            247100.0   
696       387.0       161.0         3.9063            178400.0   

    ocean_proximity  ocean_proximity_le  
290        NEAR BAY                   3  
341        NEAR BAY                   3  
538        NEAR BAY                   3  
563        NEAR BAY                   3  
696        NEAR BAY                   3

### Первый способ решить эту проблему
#### Удалить строки с нулевыми значениями

In [103]:
sample_incomplete_rows.dropna(subset=['total_bedrooms'])

Empty DataFrame
Columns: [longitude, latitude, housing_median_age, total_rooms, total_bedrooms, population, households, median_income, median_house_value, ocean_proximity, ocean_proximity_le]
Index: []

### Второй способ
#### Удалить столбцы, у которых есть нулевые значения(пропуски)

In [104]:
sample_incomplete_rows.drop("total_bedrooms", axis=1)

longitude  latitude  housing_median_age  total_rooms  population  \
290    -122.16     37.77                47.0       1256.0       570.0   
341    -122.17     37.75                38.0        992.0       732.0   
538    -122.28     37.78                29.0       5154.0      3741.0   
563    -122.24     37.75                45.0        891.0       384.0   
696    -122.10     37.69                41.0        746.0       387.0   

     households  median_income  median_house_value ocean_proximity  \
290       218.0         4.3750            161900.0        NEAR BAY   
341       259.0         1.6196             85100.0        NEAR BAY   
538      1273.0         2.5762            173400.0        NEAR BAY   
563       146.0         4.9489            247100.0        NEAR BAY   
696       161.0         3.9063            178400.0        NEAR BAY   

     ocean_proximity_le  
290                   3  
341                   3  
538                   3  
563                   3  
696                   3

### Третий способ
#### Заменить нулевые (пустые) значения средним/медианой/самой частой величиной

In [106]:
mean_ = housing['total_bedrooms'].mean()
sample_incomplete_rows['total_bedrooms'].fillna(mean_, inplace=True)
sample_incomplete_rows

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
290    -122.16     37.77                47.0       1256.0      537.870553   
341    -122.17     37.75                38.0        992.0      537.870553   
538    -122.28     37.78                29.0       5154.0      537.870553   
563    -122.24     37.75                45.0        891.0      537.870553   
696    -122.10     37.69                41.0        746.0      537.870553   

     population  households  median_income  median_house_value  \
290       570.0       218.0         4.3750            161900.0   
341       732.0       259.0         1.6196             85100.0   
538      3741.0      1273.0         2.5762            173400.0   
563       384.0       146.0         4.9489            247100.0   
696       387.0       161.0         3.9063            178400.0   

    ocean_proximity  ocean_proximity_le  
290        NEAR BAY                   3  
341        NEAR BAY                   3  
538        NEAR BAY                   3  
563        NEAR BAY                   3  
696        NEAR BAY                   3

In [107]:
median = housing['total_bedrooms'].median()
sample_incomplete_rows['total_bedrooms'].fillna(median, inplace=True)
sample_incomplete_rows

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
290    -122.16     37.77                47.0       1256.0      537.870553   
341    -122.17     37.75                38.0        992.0      537.870553   
538    -122.28     37.78                29.0       5154.0      537.870553   
563    -122.24     37.75                45.0        891.0      537.870553   
696    -122.10     37.69                41.0        746.0      537.870553   

     population  households  median_income  median_house_value  \
290       570.0       218.0         4.3750            161900.0   
341       732.0       259.0         1.6196             85100.0   
538      3741.0      1273.0         2.5762            173400.0   
563       384.0       146.0         4.9489            247100.0   
696       387.0       161.0         3.9063            178400.0   

    ocean_proximity  ocean_proximity_le  
290        NEAR BAY                   3  
341        NEAR BAY                   3  
538        NEAR BAY                   3  
563        NEAR BAY                   3  
696        NEAR BAY                   3

## Масштабирование данных

In [115]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
housing_1 = housing.copy()
housing_1.drop(['ocean_proximity'], axis=1, inplace=True)
housing_1 = scaler.fit_transform(housing_1)
df = pd.DataFrame(housing_1)
df

0         1         2         3         4         5         6  \
0     -1.327835  1.052548  0.982143 -0.804819 -0.970325 -0.974429 -0.977033   
1     -1.322844  1.043185 -0.607019  2.045890  1.348276  0.861439  1.669961   
2     -1.332827  1.038503  1.856182 -0.535746 -0.825561 -0.820777 -0.843637   
3     -1.337818  1.038503  1.856182 -0.624215 -0.718768 -0.766028 -0.733781   
4     -1.337818  1.038503  1.856182 -0.462404 -0.611974 -0.759847 -0.629157   
...         ...       ...       ...       ...       ...       ...       ...   
20635 -0.758826  1.801647 -0.289187 -0.444985 -0.388895 -0.512592 -0.443449   
20636 -0.818722  1.806329 -0.845393 -0.888704 -0.920488 -0.944405 -1.008420   
20637 -0.823713  1.778237 -0.924851 -0.174995 -0.125472 -0.369537 -0.174042   
20638 -0.873626  1.778237 -0.845393 -0.355600 -0.305834 -0.604429 -0.393753   
20639 -0.833696  1.750146 -1.004309  0.068408  0.185416 -0.033977  0.079672   

              7         8         9  
0      2.344766  2.129631  1.291089  
1      2.332238  1.314156  1.291089  
2      1.782699  1.258693  1.291089  
3      0.932968  1.165100  1.291089  
4     -0.012881  1.172900  1.291089  
...         ...       ...       ...  
20635 -1.216128 -1.115804 -0.116739  
20636 -0.691593 -1.124470 -0.116739  
20637 -1.142593 -0.992746 -0.116739  
20638 -1.054583 -1.058608 -0.116739  
20639 -0.780129 -1.017878 -0.116739  

[20640 rows x 10 columns]